### DISCUSSION TITLE
Select vs Compound Select

### DISCUSSION THREAD

**Discussion by liverpool1026** [sqlalchemy/sqlalchemy#10692] [Usage Questions] [answered]
Hi,

Migrating from the old syntax to the new syntax and I am coming across an issue on .union.

Previously, we have a method called get_query.

```
def get_query() -> sa.orm.Query:
   ....
```

and now we are converting it to 
```
def get_statement() -> sa.sql.Select:
   ...
```

But then hitting some issues on the typing. As the old .union_all still returns the same Query type. While the new syntax .union_all returns CompoundSelect instead of  Select.

We are thinking of changing the return type to 
```
def get_statement() -> sa.sql.SelectBase:
   ...
```

But am not sure if there is anything in Select which is not covered in CompoundSelect or vice versa.

Any insight on what's the best way to handle this?

**Comment by zzzeek** [comment, marked as answer]
the ORM Query when you do the UNION thing is always going to be SELECTing from that UNION as a subquery, which is why you can continue to call methods like `filter()`, `order_by()` etc. on that `Query`.  if you wanted to do the equivalent thing for `Select` you would continue to return a `Select` in all cases.  When you create a `UNION`, you'd `Select` from it, and if you are returning ORM entities you'd need to do this anyway, using one of the two forms documented at https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#selecting-entities-from-unions-and-other-set-operations , that is, `from_statement()` or `aliased()`.
